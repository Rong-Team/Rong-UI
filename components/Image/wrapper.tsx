import styled, { css } from "styled-components";

const habs=css`
    position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display:flex;
  justify-content:center;
  align-items:center;
  flex-direction:column;
`

export const ImgBase=styled.div`
  display: inline-block;
  position: relative;

`

export const ImgPlaceholder=styled.div`
    ${habs};
    background:#f0f0f0;
    `

export const ImgWrap=styled.img<{place?:boolean}>`
width: 100%;
    height: auto;
${props=>props.place?css`
    background-color: #ebebeb ;
      background-repeat: no-repeat;
      background-position: center center;
`:null}
`