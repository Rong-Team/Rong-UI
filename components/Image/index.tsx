import React, { useState } from 'react'
import { ImgBase, ImgPlaceholder, ImgWrap } from './wrapper';
import { PictureOutlined, LoadingOutlined } from '@ant-design/icons';
export interface ImageProps
  extends Omit<React.ImgHTMLAttributes<HTMLImageElement>, 'placeholder' | 'onClick'> {
  // Original
  src?: string;

  placeholder?: React.ReactNode;
  fallback?: string;
 // preview?: boolean | ImagePreviewType;
 // onPreviewClose?: (value: boolean, prevValue: boolean) => void;
  //onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
 // getPopupContainer?: () => HTMLElement;
}
type ImageStatus = 'normal' | 'error' | 'loading';

const Image:React.FC<ImageProps>=({

    src,
  alt,
//   onPreviewClose: onInitialPreviewClose,
//   getPopupContainer,

  placeholder,
  fallback,
  width="120px",
  height="120px",
  style,
//   preview = true,
  className: originalClassName,
 // onClick,

  // Img
  crossOrigin,
  decoding,
  loading,
  referrerPolicy,
  sizes,
  srcSet,
  useMap,
  ...otherProps
})=>{
    const isCustomPlaceholder = placeholder && placeholder !== true;

    const [status, setStatus] = useState<ImageStatus>( 'loading');
    const [url, seturl] = useState(src)

    const imgCommonProps = {
        crossOrigin,
        decoding,
        loading,
        referrerPolicy,
        sizes,
        srcSet,
        useMap,
        alt,
        style: height !== undefined ? { height } : undefined,
      };
    
    const onLoad = () => {
        setStatus('normal');
      };
    
      const onError = () => {
        setStatus('error');
      };
    
      const reload=()=>{
        setStatus('normal')
        seturl(prev=>prev+`?time=${Math.random()}`)
      }
    return (
        <>
          <ImgBase
            {...otherProps}
            id='rong-image'
            style={{
              ...style,
              width,
              height,
            }}
          >
            {status === 'error' ? (fallback ? (
              <ImgWrap id="fallback-img" place={placeholder?true:false}  {...imgCommonProps} src={fallback} />
            ) : (
              <ImgPlaceholder id="default-fallback" onClick={reload}>
                <PictureOutlined />
                Reload
              </ImgPlaceholder>
            )):<ImgWrap id="image-base"  place={placeholder?true:false} {...imgCommonProps} onLoad={onLoad} onError={onError} src={url} />
            }

            {status === 'loading' && (
              <ImgPlaceholder id="loading-image" aria-hidden="true">
                {placeholder||<LoadingOutlined />}
              </ImgPlaceholder>
            )}
          </ImgBase>
        </>
      );
}

export default Image