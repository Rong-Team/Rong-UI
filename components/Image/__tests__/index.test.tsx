import React from 'react'
import 'jest-styled-components'
import {mount} from 'enzyme'
import Image from '../index'
import renderer from 'react-test-renderer'
describe('Image Test', () => {
    it("should render image",()=>{
        const wrapper=mount(<Image src="https://images.unsplash.com/photo-1602524206684-fdf6393c7d89?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2850&q=80"/>)
        expect(wrapper.find("#image-base")).toBeTruthy()
    })

    it("should render fallback image when error",()=>{
        const wrapper=mount(<Image fallback="https://images.unsplash.com/photo-1602524206684-fdf6393c7d89?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2850&q=80" 
        src="error"/>)
        expect(wrapper.find("#default-fallback")).toBeTruthy()

    })

    it("should have snapshot",()=>{
        const wrapper=renderer.create(<Image src="https://images.unsplash.com/photo-1602524206684-fdf6393c7d89?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2850&q=8"/>).toJSON()
        expect(wrapper).toMatchSnapshot()
    })
    
})
