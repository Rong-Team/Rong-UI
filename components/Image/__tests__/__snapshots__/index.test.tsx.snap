// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Test should have snapshot 1`] = `
.c0 {
  display: inline-block;
  position: relative;
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background: #f0f0f0;
}

.c1 {
  width: 100%;
  height: auto;
}

<div
  className="c0"
  id="rong-image"
  style={
    Object {
      "height": "120px",
      "width": "120px",
    }
  }
>
  <img
    className="c1"
    id="image-base"
    onError={[Function]}
    onLoad={[Function]}
    src="https://images.unsplash.com/photo-1602524206684-fdf6393c7d89?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2850&q=8"
    style={
      Object {
        "height": "120px",
      }
    }
  />
  <div
    aria-hidden="true"
    className="c2"
    id="loading-image"
  >
    <span
      aria-label="loading"
      className="anticon anticon-loading"
      role="img"
    >
      <svg
        aria-hidden="true"
        className="anticon-spin"
        data-icon="loading"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
  </div>
</div>
`;
